esp32: testing.ztest.error_hook.no_userspace build fails due to array-bounds warnings

**Describe the bug**

`testing.ztest.error_hook.no_userspace` test build fails on `esp32` due to `Warray-bounds` warnings.

**To Reproduce**

```
scripts/twister -v -N -M -p esp32 -s testing.ztest.error_hook.no_userspace
```

**Expected behavior**

Test build succeeds.

**Logs and console output**

```
[81/183] Building C object zephyr/CMakeFiles/zephyr.dir/__w/zephyr/modules/hal/espressif/components/driver/periph_ctrl.c.obj
FAILED: zephyr/CMakeFiles/zephyr.dir/__w/zephyr/modules/hal/espressif/components/driver/periph_ctrl.c.obj 
ccache /opt/toolchains/zephyr-sdk-0.15.0/xtensa-espressif_esp32_zephyr-elf/bin/xtensa-espressif_esp32_zephyr-elf-gcc -DCONFIG_SPI_FLASH_ROM_DRIVER_PATCH -DCONFIG_SPI_FLASH_USE_LEGACY_IMPL -DKERNEL -DTC_RUNID=97bcfa949f4b55bbca220962305b3014 -D_FORTIFY_SOURCE=2 -D__ZEPHYR__=1 -I../../../../../../kernel/include -I../../../../../../arch/xtensa/include -I../../../../../../include -Izephyr/include/generated -I../../../../../../soc/xtensa/esp32 -I../../../../../../soc/xtensa/esp32/include -I../../../../../../subsys/testsuite/include -I../../../../../../subsys/testsuite/ztest/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/include/bt -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/include/crypto -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../esp_shared/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../esp_shared/include/wifi -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/hal/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/hal/platform_port/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/hal/esp32/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/esp_hw_support/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/esp_hw_support/port/esp32 -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/esp_hw_support/port/esp32/private_include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/esp_common/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32/ld -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/esp32/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/soc/esp32/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/xtensa/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/xtensa/esp32/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/soc/include/soc -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/soc/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/soc/src/esp32/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/driver/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/esp_wifi/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/efuse/esp32/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/efuse/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/esp_system/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/esp_wifi/esp32/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/esp_timer/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/esp_timer/private_include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/esp_netif/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/esp_event/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/log/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/wpa_supplicant/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/wpa_supplicant/port/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/wpa_supplicant/src -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/wpa_supplicant/include/esp_supplicant -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/wpa_supplicant/src/crypto -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/spi_flash/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/spi_flash/private_include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/bootloader_support/include -I/__w/zephyr/modules/hal/espressif/zephyr/esp32/../../components/bootloader_support/include_bootloader -I../../../../../../tests/ztest/error_hook/include -isystem ../../../../../../lib/libc/minimal/include -fno-strict-aliasing -Os -imacros /__w/zephyr/zephyr/twister-out/esp32/tests/ztest/error_hook/testing.ztest.error_hook.no_userspace/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always --sysroot=/opt/toolchains/zephyr-sdk-0.15.0/xtensa-espressif_esp32_zephyr-elf/xtensa-espressif_esp32_zephyr-elf -imacros /__w/zephyr/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -Werror -fno-asynchronous-unwind-tables -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/__w/zephyr/zephyr/tests/ztest/error_hook=CMAKE_SOURCE_DIR -fmacro-prefix-map=/__w/zephyr/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/__w/zephyr=WEST_TOPDIR -ffunction-sections -fdata-sections -mlongcalls -Wno-unused-variable -Wno-maybe-uninitialized -std=c99 -MD -MT zephyr/CMakeFiles/zephyr.dir/__w/zephyr/modules/hal/espressif/components/driver/periph_ctrl.c.obj -MF zephyr/CMakeFiles/zephyr.dir/__w/zephyr/modules/hal/espressif/components/driver/periph_ctrl.c.obj.d -o zephyr/CMakeFiles/zephyr.dir/__w/zephyr/modules/hal/espressif/components/driver/periph_ctrl.c.obj -c /__w/zephyr/modules/hal/espressif/components/driver/periph_ctrl.c
/__w/zephyr/modules/hal/espressif/components/driver/periph_ctrl.c: In function 'periph_module_enable':
/__w/zephyr/modules/hal/espressif/components/driver/periph_ctrl.c:28:19: error: array subscript 35 is above array bounds of 'uint8_t[35]' {aka 'unsigned char[35]'} [-Werror=array-bounds]
   28 |     if (ref_counts[periph] == 0) {
      |         ~~~~~~~~~~^~~~~~~~
/__w/zephyr/modules/hal/espressif/components/driver/periph_ctrl.c:22:16: note: while referencing 'ref_counts'
   22 | static uint8_t ref_counts[PERIPH_MODULE_MAX] = {0};
      |                ^~~~~~~~~~
/__w/zephyr/modules/hal/espressif/components/driver/periph_ctrl.c: In function 'periph_module_disable':
/__w/zephyr/modules/hal/espressif/components/driver/periph_ctrl.c:39:15: error: array subscript 35 is above array bounds of 'uint8_t[35]' {aka 'unsigned char[35]'} [-Werror=array-bounds]
   39 |     ref_counts[periph]--;
      |     ~~~~~~~~~~^~~~~~~~
/__w/zephyr/modules/hal/espressif/components/driver/periph_ctrl.c:22:16: note: while referencing 'ref_counts'
   22 | static uint8_t ref_counts[PERIPH_MODULE_MAX] = {0};
      |                ^~~~~~~~~~
/__w/zephyr/modules/hal/espressif/components/driver/periph_ctrl.c:39:15: error: array subscript 35 is above array bounds of 'uint8_t[35]' {aka 'unsigned char[35]'} [-Werror=array-bounds]
   39 |     ref_counts[periph]--;
      |     ~~~~~~~~~~^~~~~~~~
/__w/zephyr/modules/hal/espressif/components/driver/periph_ctrl.c:22:16: note: while referencing 'ref_counts'
   22 | static uint8_t ref_counts[PERIPH_MODULE_MAX] = {0};
      |                ^~~~~~~~~~
cc1: all warnings being treated as errors
```

**Environment (please complete the following information):**
 - OS: Ubuntu 20.04
 - Toolchain: Zephyr SDK 0.15.0
 - Commit SHA: e6d7118f0ee7fbf542124ab6845ba6879de8aaf7

--- a/boards/common/esp32.board.cmake
+++ b/boards/common/esp32.board.cmake
@@ -10,7 +10,7 @@ board_runner_args(openocd  --gdb-init "flushregs")
 board_runner_args(openocd  --gdb-init "mon reset halt")
 board_runner_args(openocd  --gdb-init "thb main")
 
-set(ESP_IDF_PATH ${ZEPHYR_ESPRESSIF_MODULE_DIR})
+set(ESP_IDF_PATH ${ZEPHYR_HAL_ESPRESSIF_MODULE_DIR})
 assert(ESP_IDF_PATH "ESP_IDF_PATH is not set")
 
 board_finalize_runner_args(esp32 "--esp-idf-path=${ESP_IDF_PATH}")

--- a/boards/xtensa/esp32/doc/index.rst
+++ b/boards/xtensa/esp32/doc/index.rst
@@ -47,12 +47,12 @@ System requirements
 Prerequisites
 -------------
 
-Espressif HAL requires binary blobs in order work. The west extension below performs the required
-syncronization to clone, checkout and pull the submodules:
+Espressif HAL requires WiFi and Bluetooth binary blobs in order work. Run the command
+below to retrieve those files.
 
 .. code-block:: console
 
-   west espressif update
+   west blobs fetch hal_espressif
 
 .. note::
 

--- a/boards/xtensa/esp32_net/doc/index.rst
+++ b/boards/xtensa/esp32_net/doc/index.rst
@@ -16,12 +16,12 @@ System requirements
 Prerequisites
 -------------
 
-Espressif HAL requires binary blobs in order work. The west extension below performs the required
-syncronization to clone, checkout and pull the submodules:
+Espressif HAL requires WiFi and Bluetooth binary blobs in order work. Run the command
+below to retrieve those files.
 
 .. code-block:: console
 
-   west espressif update
+   west blobs fetch hal_espressif
 
 .. note::
 

--- a/boards/xtensa/esp32s2_saola/doc/index.rst
+++ b/boards/xtensa/esp32s2_saola/doc/index.rst
@@ -36,12 +36,12 @@ System requirements
 Prerequisites
 -------------
 
-Espressif HAL requires binary blobs in order work. The west extension below performs the required
-syncronization to clone, checkout and pull the submodules:
+Espressif HAL requires WiFi and Bluetooth binary blobs in order work. Run the command
+below to retrieve those files.
 
 .. code-block:: console
 
-   west espressif update
+   west blobs fetch hal_espressif
 
 .. note::
 

--- a/boards/xtensa/esp_wrover_kit/doc/index.rst
+++ b/boards/xtensa/esp_wrover_kit/doc/index.rst
@@ -445,12 +445,12 @@ System requirements
 Prerequisites
 -------------
 
-Espressif HAL requires binary blobs in order work. The west extension below performs the required
-syncronization to clone, checkout and pull the submodules:
+Espressif HAL requires WiFi and Bluetooth binary blobs in order work. Run the command
+below to retrieve those files.
 
 .. code-block:: console
 
-   west espressif update
+   west blobs fetch hal_espressif
 
 .. note::
 

--- a/boards/xtensa/heltec_wifi_lora32_v2/doc/index.rst
+++ b/boards/xtensa/heltec_wifi_lora32_v2/doc/index.rst
@@ -28,12 +28,12 @@ System requirements
 Prerequisites
 -------------
 
-Espressif HAL requires binary blobs in order work. The west extension below performs the required
-syncronization to clone, checkout and pull the submodules:
+Espressif HAL requires WiFi and Bluetooth binary blobs in order work. Run the command
+below to retrieve those files.
 
 .. code-block:: console
 
-   west espressif update
+   west blobs fetch hal_espressif
 
 .. note::
 

--- a/boards/xtensa/odroid_go/doc/index.rst
+++ b/boards/xtensa/odroid_go/doc/index.rst
@@ -84,12 +84,12 @@ System requirements
 Prerequisites
 -------------
 
-Espressif HAL requires binary blobs in order work. The west extension below performs the required
-syncronization to clone, checkout and pull the submodules:
+Espressif HAL requires WiFi and Bluetooth binary blobs in order work. Run the command
+below to retrieve those files.
 
 .. code-block:: console
 
-   west espressif update
+   west blobs fetch hal_espressif
 
 .. note::
 

--- a/boards/xtensa/olimex_esp32_evb/doc/index.rst
+++ b/boards/xtensa/olimex_esp32_evb/doc/index.rst
@@ -94,12 +94,12 @@ System requirements
 Prerequisites
 -------------
 
-Espressif HAL requires binary blobs in order work. The west extension below performs the required
-syncronization to clone, checkout and pull the submodules:
+Espressif HAL requires WiFi and Bluetooth binary blobs in order work. Run the command
+below to retrieve those files.
 
 .. code-block:: console
 
-   west espressif update
+   west blobs fetch hal_espressif
 
 .. note::
 

--- a/drivers/clock_control/clock_control_esp32.c
+++ b/drivers/clock_control/clock_control_esp32.c
@@ -20,6 +20,7 @@
 #define DT_CPU_COMPAT cdns_tensilica_xtensa_lx7
 #include <zephyr/dt-bindings/clock/esp32s2_clock.h>
 #include "esp32s2/rom/rtc.h"
+#include "soc/dport_reg.h"
 #elif CONFIG_IDF_TARGET_ESP32C3
 #define DT_CPU_COMPAT espressif_riscv
 #include <zephyr/dt-bindings/clock/esp32c3_clock.h>
@@ -29,7 +30,9 @@
 #include <soc/rtc.h>
 #endif
 
+#include "esp_rom_sys.h"
 #include <soc/rtc.h>
+#include <soc/i2s_reg.h>
 #include <soc/apb_ctrl_reg.h>
 #include <soc/timer_group_reg.h>
 #include <hal/clk_gate_ll.h>
@@ -115,6 +118,294 @@ static int clock_control_esp32_get_rate(const struct device *dev,
 	return 0;
 }
 
+#if defined(CONFIG_SOC_ESP32) || defined(CONFIG_SOC_ESP32_NET)
+static void esp32_clock_perip_init(void)
+{
+	uint32_t common_perip_clk;
+	uint32_t hwcrypto_perip_clk;
+	uint32_t wifi_bt_sdio_clk;
+
+#if !CONFIG_SMP
+	soc_reset_reason_t rst_reas[1];
+#else
+	soc_reset_reason_t rst_reas[2];
+#endif
+
+	rst_reas[0] = esp_rom_get_reset_reason(0);
+#if CONFIG_SMP
+	rst_reas[1] = esp_rom_get_reset_reason(1);
+#endif
+
+	/* For reason that only reset CPU, do not disable the clocks
+	 * that have been enabled before reset.
+	 */
+	if ((rst_reas[0] == RESET_REASON_CPU0_MWDT0 || rst_reas[0] == RESET_REASON_CPU0_SW ||
+		rst_reas[0] == RESET_REASON_CPU0_RTC_WDT)
+#if CONFIG_SMP
+		|| (rst_reas[1] == RESET_REASON_CPU1_MWDT1 || rst_reas[1] == RESET_REASON_CPU1_SW ||
+			rst_reas[1] == RESET_REASON_CPU1_RTC_WDT)
+#endif
+	) {
+		common_perip_clk = ~DPORT_READ_PERI_REG(DPORT_PERIP_CLK_EN_REG);
+		hwcrypto_perip_clk = ~DPORT_READ_PERI_REG(DPORT_PERI_CLK_EN_REG);
+		wifi_bt_sdio_clk = ~DPORT_READ_PERI_REG(DPORT_WIFI_CLK_EN_REG);
+	} else {
+		common_perip_clk = DPORT_WDG_CLK_EN |
+			DPORT_PCNT_CLK_EN |
+			DPORT_LEDC_CLK_EN |
+			DPORT_TIMERGROUP1_CLK_EN |
+			DPORT_PWM0_CLK_EN |
+			DPORT_TWAI_CLK_EN |
+			DPORT_PWM1_CLK_EN |
+			DPORT_PWM2_CLK_EN |
+			DPORT_PWM3_CLK_EN;
+
+		hwcrypto_perip_clk = DPORT_PERI_EN_AES |
+			DPORT_PERI_EN_SHA |
+			DPORT_PERI_EN_RSA |
+			DPORT_PERI_EN_SECUREBOOT;
+
+		wifi_bt_sdio_clk = DPORT_WIFI_CLK_WIFI_EN |
+			DPORT_WIFI_CLK_BT_EN_M |
+			DPORT_WIFI_CLK_UNUSED_BIT5 |
+			DPORT_WIFI_CLK_UNUSED_BIT12 |
+			DPORT_WIFI_CLK_SDIOSLAVE_EN |
+			DPORT_WIFI_CLK_SDIO_HOST_EN |
+			DPORT_WIFI_CLK_EMAC_EN;
+	}
+
+	/* Reset peripherals like I2C, SPI, UART, I2S and bring them to known state */
+	common_perip_clk |= DPORT_I2S0_CLK_EN |
+			DPORT_UART_CLK_EN |
+			DPORT_SPI2_CLK_EN |
+			DPORT_I2C_EXT0_CLK_EN |
+			DPORT_UHCI0_CLK_EN |
+			DPORT_RMT_CLK_EN |
+			DPORT_UHCI1_CLK_EN |
+			DPORT_SPI3_CLK_EN |
+			DPORT_I2C_EXT1_CLK_EN |
+			DPORT_I2S1_CLK_EN |
+			DPORT_SPI_DMA_CLK_EN;
+
+	common_perip_clk &= ~DPORT_SPI01_CLK_EN;
+	common_perip_clk &= ~DPORT_SPI2_CLK_EN;
+	common_perip_clk &= ~DPORT_SPI3_CLK_EN;
+
+	/* Change I2S clock to audio PLL first. Because if I2S uses 160MHz clock,
+	 * the current is not reduced when disable I2S clock.
+	 */
+	DPORT_SET_PERI_REG_MASK(I2S_CLKM_CONF_REG(0), I2S_CLKA_ENA);
+	DPORT_SET_PERI_REG_MASK(I2S_CLKM_CONF_REG(1), I2S_CLKA_ENA);
+
+	/* Disable some peripheral clocks. */
+	DPORT_CLEAR_PERI_REG_MASK(DPORT_PERIP_CLK_EN_REG, common_perip_clk);
+	DPORT_SET_PERI_REG_MASK(DPORT_PERIP_RST_EN_REG, common_perip_clk);
+
+	/* Disable hardware crypto clocks. */
+	DPORT_CLEAR_PERI_REG_MASK(DPORT_PERI_CLK_EN_REG, hwcrypto_perip_clk);
+	DPORT_SET_PERI_REG_MASK(DPORT_PERI_RST_EN_REG, hwcrypto_perip_clk);
+
+	/* Disable WiFi/BT/SDIO clocks. */
+	DPORT_CLEAR_PERI_REG_MASK(DPORT_WIFI_CLK_EN_REG, wifi_bt_sdio_clk);
+
+	/* Enable RNG clock. */
+	periph_module_enable(PERIPH_RNG_MODULE);
+}
+#endif
+
+#if defined(CONFIG_SOC_ESP32S2)
+static void esp32_clock_perip_init(void)
+{
+	uint32_t common_perip_clk;
+	uint32_t hwcrypto_perip_clk;
+	uint32_t wifi_bt_sdio_clk;
+	uint32_t common_perip_clk1;
+
+	soc_reset_reason_t rst_reason = esp_rom_get_reset_reason(0);
+
+	/* For reason that only reset CPU, do not disable the clocks
+	 * that have been enabled before reset.
+	 */
+	if (rst_reason == RESET_REASON_CPU0_MWDT0 || rst_reason == RESET_REASON_CPU0_SW ||
+		rst_reason == RESET_REASON_CPU0_RTC_WDT || rst_reason == RESET_REASON_CPU0_MWDT1) {
+		common_perip_clk = ~DPORT_READ_PERI_REG(DPORT_PERIP_CLK_EN_REG);
+		hwcrypto_perip_clk = ~DPORT_READ_PERI_REG(DPORT_PERIP_CLK_EN1_REG);
+		wifi_bt_sdio_clk = ~DPORT_READ_PERI_REG(DPORT_WIFI_CLK_EN_REG);
+	} else {
+		common_perip_clk = DPORT_WDG_CLK_EN |
+			DPORT_I2S0_CLK_EN |
+			DPORT_UART1_CLK_EN |
+			DPORT_SPI2_CLK_EN |
+			DPORT_I2C_EXT0_CLK_EN |
+			DPORT_UHCI0_CLK_EN |
+			DPORT_RMT_CLK_EN |
+			DPORT_PCNT_CLK_EN |
+			DPORT_LEDC_CLK_EN |
+			DPORT_TIMERGROUP1_CLK_EN |
+			DPORT_SPI3_CLK_EN |
+			DPORT_PWM0_CLK_EN |
+			DPORT_TWAI_CLK_EN |
+			DPORT_PWM1_CLK_EN |
+			DPORT_I2S1_CLK_EN |
+			DPORT_SPI2_DMA_CLK_EN |
+			DPORT_SPI3_DMA_CLK_EN |
+			DPORT_PWM2_CLK_EN |
+			DPORT_PWM3_CLK_EN;
+
+		common_perip_clk1 = 0;
+
+		hwcrypto_perip_clk = DPORT_CRYPTO_AES_CLK_EN |
+				DPORT_CRYPTO_SHA_CLK_EN |
+				DPORT_CRYPTO_RSA_CLK_EN;
+
+		wifi_bt_sdio_clk = DPORT_WIFI_CLK_WIFI_EN |
+			DPORT_WIFI_CLK_BT_EN_M |
+			DPORT_WIFI_CLK_UNUSED_BIT5 |
+			DPORT_WIFI_CLK_UNUSED_BIT12 |
+			DPORT_WIFI_CLK_SDIOSLAVE_EN |
+			DPORT_WIFI_CLK_SDIO_HOST_EN |
+			DPORT_WIFI_CLK_EMAC_EN;
+	}
+
+	/* Reset peripherals like I2C, SPI, UART, I2S and bring them to known state */
+	common_perip_clk |= DPORT_I2S0_CLK_EN |
+			DPORT_UART1_CLK_EN |
+			DPORT_USB_CLK_EN |
+			DPORT_SPI2_CLK_EN |
+			DPORT_I2C_EXT0_CLK_EN |
+			DPORT_UHCI0_CLK_EN |
+			DPORT_RMT_CLK_EN |
+			DPORT_UHCI1_CLK_EN |
+			DPORT_SPI3_CLK_EN |
+			DPORT_I2C_EXT1_CLK_EN |
+			DPORT_I2S1_CLK_EN |
+			DPORT_SPI2_DMA_CLK_EN |
+			DPORT_SPI3_DMA_CLK_EN;
+
+	common_perip_clk1 = 0;
+
+	/* Change I2S clock to audio PLL first. Because if I2S uses 160MHz clock,
+	 * the current is not reduced when disable I2S clock.
+	 */
+	REG_SET_FIELD(I2S_CLKM_CONF_REG(0), I2S_CLK_SEL, I2S_CLK_AUDIO_PLL);
+	REG_SET_FIELD(I2S_CLKM_CONF_REG(1), I2S_CLK_SEL, I2S_CLK_AUDIO_PLL);
+
+	/* Disable some peripheral clocks. */
+	DPORT_CLEAR_PERI_REG_MASK(DPORT_PERIP_CLK_EN_REG, common_perip_clk);
+	DPORT_SET_PERI_REG_MASK(DPORT_PERIP_RST_EN_REG, common_perip_clk);
+
+	DPORT_CLEAR_PERI_REG_MASK(DPORT_PERIP_CLK_EN1_REG, common_perip_clk1);
+	DPORT_SET_PERI_REG_MASK(DPORT_PERIP_RST_EN1_REG, common_perip_clk1);
+
+	/* Disable hardware crypto clocks. */
+	DPORT_CLEAR_PERI_REG_MASK(DPORT_PERIP_CLK_EN1_REG, hwcrypto_perip_clk);
+	DPORT_SET_PERI_REG_MASK(DPORT_PERIP_RST_EN1_REG, hwcrypto_perip_clk);
+
+	/* Disable WiFi/BT/SDIO clocks. */
+	DPORT_CLEAR_PERI_REG_MASK(DPORT_WIFI_CLK_EN_REG, wifi_bt_sdio_clk);
+
+	/* Enable WiFi MAC and POWER clocks */
+	DPORT_SET_PERI_REG_MASK(DPORT_WIFI_CLK_EN_REG, DPORT_WIFI_CLK_WIFI_EN);
+
+	/* Set WiFi light sleep clock source to RTC slow clock */
+	DPORT_REG_SET_FIELD(DPORT_BT_LPCK_DIV_INT_REG, DPORT_BT_LPCK_DIV_NUM, 0);
+	DPORT_CLEAR_PERI_REG_MASK(DPORT_BT_LPCK_DIV_FRAC_REG, DPORT_LPCLK_SEL_8M);
+	DPORT_SET_PERI_REG_MASK(DPORT_BT_LPCK_DIV_FRAC_REG, DPORT_LPCLK_SEL_RTC_SLOW);
+
+	/* Enable RNG clock. */
+	periph_module_enable(PERIPH_RNG_MODULE);
+}
+#endif
+
+#if defined(CONFIG_SOC_ESP32C3)
+static void esp32_clock_perip_init(void)
+{
+	uint32_t common_perip_clk;
+	uint32_t hwcrypto_perip_clk;
+	uint32_t wifi_bt_sdio_clk;
+	uint32_t common_perip_clk1;
+
+	soc_reset_reason_t rst_reason = esp_rom_get_reset_reason(0);
+
+	/* For reason that only reset CPU, do not disable the clocks
+	 * that have been enabled before reset.
+	 */
+	if (rst_reason == RESET_REASON_CPU0_MWDT0 || rst_reason == RESET_REASON_CPU0_SW ||
+		rst_reason == RESET_REASON_CPU0_RTC_WDT || rst_reason == RESET_REASON_CPU0_MWDT1) {
+		common_perip_clk = ~READ_PERI_REG(SYSTEM_PERIP_CLK_EN0_REG);
+		hwcrypto_perip_clk = ~READ_PERI_REG(SYSTEM_PERIP_CLK_EN1_REG);
+		wifi_bt_sdio_clk = ~READ_PERI_REG(SYSTEM_WIFI_CLK_EN_REG);
+	} else {
+		common_perip_clk = SYSTEM_WDG_CLK_EN |
+				SYSTEM_I2S0_CLK_EN |
+				SYSTEM_UART1_CLK_EN |
+				SYSTEM_SPI2_CLK_EN |
+				SYSTEM_I2C_EXT0_CLK_EN |
+				SYSTEM_UHCI0_CLK_EN |
+				SYSTEM_RMT_CLK_EN |
+				SYSTEM_LEDC_CLK_EN |
+				SYSTEM_TIMERGROUP1_CLK_EN |
+				SYSTEM_SPI3_CLK_EN |
+				SYSTEM_SPI4_CLK_EN |
+				SYSTEM_TWAI_CLK_EN |
+				SYSTEM_I2S1_CLK_EN |
+				SYSTEM_SPI2_DMA_CLK_EN |
+				SYSTEM_SPI3_DMA_CLK_EN;
+
+		common_perip_clk1 = 0;
+
+		hwcrypto_perip_clk = SYSTEM_CRYPTO_AES_CLK_EN |
+				SYSTEM_CRYPTO_SHA_CLK_EN |
+				SYSTEM_CRYPTO_RSA_CLK_EN;
+
+		wifi_bt_sdio_clk = SYSTEM_WIFI_CLK_WIFI_EN |
+				SYSTEM_WIFI_CLK_BT_EN_M |
+				SYSTEM_WIFI_CLK_UNUSED_BIT5 |
+				SYSTEM_WIFI_CLK_UNUSED_BIT12;
+	}
+
+	/* Reset peripherals like I2C, SPI, UART, I2S and bring them to known state */
+	common_perip_clk |= SYSTEM_I2S0_CLK_EN |
+			SYSTEM_UART1_CLK_EN |
+			SYSTEM_SPI2_CLK_EN |
+			SYSTEM_I2C_EXT0_CLK_EN |
+			SYSTEM_UHCI0_CLK_EN |
+			SYSTEM_RMT_CLK_EN |
+			SYSTEM_UHCI1_CLK_EN |
+			SYSTEM_SPI3_CLK_EN |
+			SYSTEM_SPI4_CLK_EN |
+			SYSTEM_I2C_EXT1_CLK_EN |
+			SYSTEM_I2S1_CLK_EN |
+			SYSTEM_SPI2_DMA_CLK_EN |
+			SYSTEM_SPI3_DMA_CLK_EN;
+
+	common_perip_clk1 = 0;
+
+	/* Disable some peripheral clocks. */
+	CLEAR_PERI_REG_MASK(SYSTEM_PERIP_CLK_EN0_REG, common_perip_clk);
+	SET_PERI_REG_MASK(SYSTEM_PERIP_RST_EN0_REG, common_perip_clk);
+
+	CLEAR_PERI_REG_MASK(SYSTEM_PERIP_CLK_EN1_REG, common_perip_clk1);
+	SET_PERI_REG_MASK(SYSTEM_PERIP_RST_EN1_REG, common_perip_clk1);
+
+	/* Disable hardware crypto clocks. */
+	CLEAR_PERI_REG_MASK(SYSTEM_PERIP_CLK_EN1_REG, hwcrypto_perip_clk);
+	SET_PERI_REG_MASK(SYSTEM_PERIP_RST_EN1_REG, hwcrypto_perip_clk);
+
+	/* Disable WiFi/BT/SDIO clocks. */
+	CLEAR_PERI_REG_MASK(SYSTEM_WIFI_CLK_EN_REG, wifi_bt_sdio_clk);
+	SET_PERI_REG_MASK(SYSTEM_WIFI_CLK_EN_REG, SYSTEM_WIFI_CLK_EN);
+
+	/* Set WiFi light sleep clock source to RTC slow clock */
+	REG_SET_FIELD(SYSTEM_BT_LPCK_DIV_INT_REG, SYSTEM_BT_LPCK_DIV_NUM, 0);
+	CLEAR_PERI_REG_MASK(SYSTEM_BT_LPCK_DIV_FRAC_REG, SYSTEM_LPCLK_SEL_8M);
+	SET_PERI_REG_MASK(SYSTEM_BT_LPCK_DIV_FRAC_REG, SYSTEM_LPCLK_SEL_RTC_SLOW);
+
+	/* Enable RNG clock. */
+	periph_module_enable(PERIPH_RNG_MODULE);
+}
+#endif
+
 static int clock_control_esp32_init(const struct device *dev)
 {
 	const struct esp32_clock_config *cfg = dev->config;
@@ -162,8 +453,7 @@ static int clock_control_esp32_init(const struct device *dev)
 	/* Re-calculate the ccount to make time calculation correct */
 	cpu_hal_set_cycle_count((uint64_t)cpu_hal_get_cycle_count() * new_freq_mhz / old_freq_mhz);
 
-	/* Enable RNG clock. */
-	periph_module_enable(PERIPH_RNG_MODULE);
+	esp32_clock_perip_init();
 
 	return 0;
 }

--- a/drivers/interrupt_controller/intc_esp32.c
+++ b/drivers/interrupt_controller/intc_esp32.c
@@ -873,25 +873,26 @@ void IRAM_ATTR esp_intr_noniram_disable(void)
 {
 	int oldint;
 	int cpu = esp_core_id();
-	int intmask = ~non_iram_int_mask[cpu];
+	int non_iram_ints = ~non_iram_int_mask[cpu];
 
 	if (non_iram_int_disabled_flag[cpu]) {
 		abort();
 	}
 	non_iram_int_disabled_flag[cpu] = true;
-	oldint = interrupt_controller_hal_disable_int_mask(intmask);
+	oldint = interrupt_controller_hal_read_interrupt_mask();
+	interrupt_controller_hal_disable_interrupts(non_iram_ints);
 	/* Save which ints we did disable */
-	non_iram_int_disabled[cpu] = oldint & non_iram_int_mask[cpu];
+	non_iram_int_disabled[cpu] = oldint & non_iram_ints;
 }
 
 void IRAM_ATTR esp_intr_noniram_enable(void)
 {
 	int cpu = esp_core_id();
-	int intmask = non_iram_int_disabled[cpu];
+	int non_iram_ints = non_iram_int_disabled[cpu];
 
 	if (!non_iram_int_disabled_flag[cpu]) {
 		abort();
 	}
 	non_iram_int_disabled_flag[cpu] = false;
-	interrupt_controller_hal_enable_int_mask(intmask);
+	interrupt_controller_hal_enable_interrupts(non_iram_ints);
 }

--- a/drivers/pwm/pwm_mc_esp32.c
+++ b/drivers/pwm/pwm_mc_esp32.c
@@ -8,6 +8,7 @@
 
 #include <hal/mcpwm_hal.h>
 #include <hal/mcpwm_ll.h>
+#include "driver/mcpwm.h"
 
 #include <soc.h>
 #include <errno.h>

--- a/drivers/spi/spi_esp32_spim.c
+++ b/drivers/spi/spi_esp32_spim.c
@@ -131,24 +131,24 @@ static int spi_esp32_init(const struct device *dev)
 	return 0;
 }
 
-static inline spi_ll_io_mode_t spi_esp32_get_io_mode(uint16_t operation)
+static inline uint8_t spi_esp32_get_line_mode(uint16_t operation)
 {
 	if (IS_ENABLED(CONFIG_SPI_EXTENDED_MODES)) {
 		switch (operation & SPI_LINES_MASK) {
 		case SPI_LINES_SINGLE:
-			return SPI_LL_IO_MODE_NORMAL;
+			return 1;
 		case SPI_LINES_DUAL:
-			return SPI_LL_IO_MODE_DUAL;
+			return 2;
 		case SPI_LINES_OCTAL:
-			return SPI_LL_IO_MODE_QIO;
+			return 8;
 		case SPI_LINES_QUAD:
-			return SPI_LL_IO_MODE_QUAD;
+			return 4;
 		default:
 			break;
 		}
 	}
 
-	return SPI_LL_IO_MODE_NORMAL;
+	return 1;
 }
 
 static int IRAM_ATTR spi_esp32_configure(const struct device *dev,
@@ -206,6 +206,7 @@ static int IRAM_ATTR spi_esp32_configure(const struct device *dev,
 		.duty_cycle = cfg->duty_cycle == 0 ? 128 : cfg->duty_cycle,
 		.input_delay_ns = cfg->input_delay_ns,
 		.use_gpio = !cfg->use_iomux,
+
 	};
 
 	spi_hal_cal_clock_conf(&timing_param, &freq, &hal_dev->timing_conf);
@@ -215,7 +216,14 @@ static int IRAM_ATTR spi_esp32_configure(const struct device *dev,
 	hal_dev->tx_lsbfirst = spi_cfg->operation & SPI_TRANSFER_LSB ? 1 : 0;
 	hal_dev->rx_lsbfirst = spi_cfg->operation & SPI_TRANSFER_LSB ? 1 : 0;
 
-	data->trans_config.io_mode = spi_esp32_get_io_mode(spi_cfg->operation);
+	data->trans_config.line_mode.data_lines = spi_esp32_get_line_mode(spi_cfg->operation);
+
+	/* multiline for command and address not supported */
+	data->trans_config.line_mode.addr_lines = 1;
+	data->trans_config.line_mode.cmd_lines = 1;
+
+	/* keep cs line after transmission not supported */
+	data->trans_config.cs_keep_active = 0;
 
 	/* SPI mode */
 	hal_dev->mode = 0;

--- a/drivers/timer/esp32c3_sys_timer.c
+++ b/drivers/timer/esp32c3_sys_timer.c
@@ -34,25 +34,30 @@ const int32_t z_sys_timer_irq_for_test = DT_IRQN(DT_NODELABEL(systimer0));
 static struct k_spinlock lock;
 static uint64_t last_count;
 
+/* Systimer HAL layer object */
+static systimer_hal_context_t systimer_hal;
+
 static void set_systimer_alarm(uint64_t time)
 {
-	systimer_hal_select_alarm_mode(SYSTIMER_ALARM_0, SYSTIMER_ALARM_MODE_ONESHOT);
-	systimer_hal_set_alarm_target(SYSTIMER_ALARM_0, time);
-	systimer_hal_enable_alarm_int(SYSTIMER_ALARM_0);
+	systimer_hal_select_alarm_mode(&systimer_hal,
+		SYSTIMER_LL_ALARM_OS_TICK_CORE0, SYSTIMER_ALARM_MODE_ONESHOT);
+
+	systimer_hal_set_alarm_target(&systimer_hal, SYSTIMER_LL_ALARM_OS_TICK_CORE0, time);
+	systimer_hal_enable_alarm_int(&systimer_hal, SYSTIMER_LL_ALARM_OS_TICK_CORE0);
 }
 
-static uint64_t systimer_alarm(void)
+static uint64_t get_systimer_alarm(void)
 {
-	return systimer_hal_get_time(SYSTIMER_COUNTER_1);
+	return systimer_hal_get_time(&systimer_hal, SYSTIMER_LL_COUNTER_OS_TICK);
 }
 
 static void sys_timer_isr(const void *arg)
 {
 	ARG_UNUSED(arg);
-	systimer_ll_clear_alarm_int(SYSTIMER_ALARM_0);
+	systimer_ll_clear_alarm_int(systimer_hal.dev, SYSTIMER_LL_ALARM_OS_TICK_CORE0);
 
 	k_spinlock_key_t key = k_spin_lock(&lock);
-	uint64_t now = systimer_alarm();
+	uint64_t now = get_systimer_alarm();
 
 	uint32_t dticks = (uint32_t)((now - last_count) / CYC_PER_TICK);
 
@@ -80,7 +85,7 @@ void sys_clock_set_timeout(int32_t ticks, bool idle)
 	ticks = CLAMP(ticks - 1, 0, (int32_t)MAX_TICKS);
 
 	k_spinlock_key_t key = k_spin_lock(&lock);
-	uint64_t now = systimer_alarm();
+	uint64_t now = get_systimer_alarm();
 	uint32_t adj, cyc = ticks * CYC_PER_TICK;
 
 	/* Round up to next tick boundary. */
@@ -108,7 +113,7 @@ uint32_t sys_clock_elapsed(void)
 	}
 
 	k_spinlock_key_t key = k_spin_lock(&lock);
-	uint32_t ret = ((uint32_t)systimer_alarm() - (uint32_t)last_count) / CYC_PER_TICK;
+	uint32_t ret = ((uint32_t)get_systimer_alarm() - (uint32_t)last_count) / CYC_PER_TICK;
 
 	k_spin_unlock(&lock, key);
 	return ret;
@@ -116,12 +121,12 @@ uint32_t sys_clock_elapsed(void)
 
 uint32_t sys_clock_cycle_get_32(void)
 {
-	return (uint32_t)systimer_alarm();
+	return (uint32_t)get_systimer_alarm();
 }
 
 uint64_t sys_clock_cycle_get_64(void)
 {
-	return systimer_alarm();
+	return get_systimer_alarm();
 }
 
 static int sys_clock_driver_init(const struct device *dev)
@@ -134,13 +139,14 @@ static int sys_clock_driver_init(const struct device *dev)
 		NULL,
 		NULL);
 
-	systimer_hal_init();
-	systimer_hal_connect_alarm_counter(SYSTIMER_ALARM_0, SYSTIMER_COUNTER_1);
-	systimer_hal_enable_counter(SYSTIMER_COUNTER_1);
-	systimer_hal_counter_can_stall_by_cpu(SYSTIMER_COUNTER_1, 0, true);
-	last_count = systimer_alarm();
-	set_systimer_alarm(last_count + CYC_PER_TICK);
+	systimer_hal_init(&systimer_hal);
+	systimer_hal_connect_alarm_counter(&systimer_hal,
+		SYSTIMER_LL_ALARM_OS_TICK_CORE0, SYSTIMER_LL_COUNTER_OS_TICK);
 
+	systimer_hal_enable_counter(&systimer_hal, SYSTIMER_LL_COUNTER_OS_TICK);
+	systimer_hal_counter_can_stall_by_cpu(&systimer_hal, SYSTIMER_LL_COUNTER_OS_TICK, 0, true);
+	last_count = get_systimer_alarm();
+	set_systimer_alarm(last_count + CYC_PER_TICK);
 	return 0;
 }
 

--- a/drivers/wifi/esp32/src/esp_wifi_drv.c
+++ b/drivers/wifi/esp32/src/esp_wifi_drv.c
@@ -197,8 +197,6 @@ static int eth_esp32_dev_init(const struct device *dev)
 
 	wifi_init_config_t config = WIFI_INIT_CONFIG_DEFAULT();
 	esp_err_t ret = esp_wifi_init(&config);
-
-	ret |= esp_supplicant_init();
 	ret |= esp_wifi_start();
 
 	if (IS_ENABLED(CONFIG_ESP32_WIFI_STA_AUTO)) {

--- a/dts/xtensa/espressif/esp32.dtsi
+++ b/dts/xtensa/espressif/esp32.dtsi
@@ -94,7 +94,7 @@
 				compatible = "soc-nv-flash";
 				reg = <0 0x400000>;
 				erase-block-size = <4096>;
-				write-block-size = <32>;
+				write-block-size = <4>;
 			};
 		};
 

--- a/include/zephyr/drivers/interrupt_controller/intc_esp32c3.h
+++ b/include/zephyr/drivers/interrupt_controller/intc_esp32c3.h
@@ -17,7 +17,28 @@
 
 /* Keep the LEVELx values as they are here; they match up with (1<<level) */
 #define ESP_INTR_FLAG_LEVEL1		(1<<1)	/* Accept a Level 1 int vector, lowest priority */
-#define ESP_INTR_FLAG_EDGE		    (1<<9)	/* Edge-triggered interrupt */
+#define ESP_INTR_FLAG_LEVEL2		(1<<2)	/* Accept a Level 2 int vector */
+#define ESP_INTR_FLAG_LEVEL3		(1<<3)	/* Accept a Level 3 int vector */
+#define ESP_INTR_FLAG_LEVEL4		(1<<4)	/* Accept a Level 4 int vector */
+#define ESP_INTR_FLAG_LEVEL5		(1<<5)	/* Accept a Level 5 int vector */
+#define ESP_INTR_FLAG_LEVEL6		(1<<6)	/* Accept a Level 6 int vector */
+#define ESP_INTR_FLAG_NMI		(1<<7)	/* Accept a Level 7 int vector, highest priority */
+#define ESP_INTR_FLAG_SHARED		(1<<8)	/* Interrupt can be shared between ISRs */
+#define ESP_INTR_FLAG_EDGE		(1<<9)	/* Edge-triggered interrupt */
+#define ESP_INTR_FLAG_IRAM		(1<<10)	/* ISR can be called if cache is disabled */
+#define ESP_INTR_FLAG_INTRDISABLED	(1<<11)	/* Return with this interrupt disabled */
+
+/* Low and medium prio interrupts. These can be handled in C. */
+#define ESP_INTR_FLAG_LOWMED	(ESP_INTR_FLAG_LEVEL1|ESP_INTR_FLAG_LEVEL2|ESP_INTR_FLAG_LEVEL3)
+
+/* High level interrupts. Need to be handled in assembly. */
+#define ESP_INTR_FLAG_HIGH	(ESP_INTR_FLAG_LEVEL4|ESP_INTR_FLAG_LEVEL5|ESP_INTR_FLAG_LEVEL6| \
+				 ESP_INTR_FLAG_NMI)
+
+/* Mask for all level flags */
+#define ESP_INTR_FLAG_LEVELMASK	(ESP_INTR_FLAG_LEVEL1|ESP_INTR_FLAG_LEVEL2|ESP_INTR_FLAG_LEVEL3| \
+				 ESP_INTR_FLAG_LEVEL4|ESP_INTR_FLAG_LEVEL5|ESP_INTR_FLAG_LEVEL6| \
+				 ESP_INTR_FLAG_NMI)
 
 /* Function prototype for interrupt handler function */
 typedef void (*isr_handler_t)(const void *arg);

--- a/soc/riscv/esp32c3/soc.c
+++ b/soc/riscv/esp32c3/soc.c
@@ -12,6 +12,8 @@
 #include <soc/system_reg.h>
 #include <soc/cache_memory.h>
 #include "hal/soc_ll.h"
+#include "esp_cpu.h"
+#include "esp_timer.h"
 #include "esp_spi_flash.h"
 #include <soc/interrupt_reg.h>
 #include <zephyr/drivers/interrupt_controller/intc_esp32c3.h>
@@ -79,6 +81,8 @@ void __attribute__((section(".iram1"))) __esp_platform_start(void)
 	REG_CLR_BIT(SYSTEM_WIFI_CLK_EN_REG, SYSTEM_WIFI_CLK_SDIOSLAVE_EN);
 	SET_PERI_REG_MASK(SYSTEM_WIFI_CLK_EN_REG, SYSTEM_WIFI_CLK_EN);
 
+	esp_timer_early_init();
+
 #if CONFIG_SOC_FLASH_ESP32
 	spi_flash_guard_set(&g_flash_guard_default_ops);
 #endif
@@ -110,7 +114,6 @@ void IRAM_ATTR esp_restart_noos(void)
 	/* Flush any data left in UART FIFOs */
 	esp_rom_uart_tx_wait_idle(0);
 	esp_rom_uart_tx_wait_idle(1);
-	esp_rom_uart_tx_wait_idle(2);
 
 	/* 2nd stage bootloader reconfigures SPI flash signals. */
 	/* Reset them to the defaults expected by ROM */
@@ -129,6 +132,9 @@ void IRAM_ATTR esp_restart_noos(void)
 			BLE_REG_REST_BIT | BLE_PWR_REG_REST_BIT | BLE_BB_REG_REST_BIT);
 
 	REG_WRITE(SYSTEM_CORE_RST_EN_REG, 0);
+
+	/* Reset uart0 core first, then reset apb side. */
+	SET_PERI_REG_MASK(UART_CLK_CONF_REG(0), UART_RST_CORE_M);
 
 	/* Reset timer/spi/uart */
 	SET_PERI_REG_MASK(SYSTEM_PERIP_RST_EN0_REG,

--- a/soc/riscv/esp32c3/soc.h
+++ b/soc/riscv/esp32c3/soc.h
@@ -12,7 +12,7 @@
 #include <rom/spi_flash.h>
 #include <zephyr/types.h>
 #include <stdbool.h>
-#include <esp_clk.h>
+#include "esp32c3/clk.h"
 #endif
 
 /* IRQ numbers */

--- a/soc/xtensa/esp32/linker.ld
+++ b/soc/xtensa/esp32/linker.ld
@@ -202,6 +202,8 @@ SECTIONS
     *(.tdata.*)
     *(.tbss)
     *(.tbss.*)
+    *(.rodata_wlog)
+    *(.rodata_wlog*)
     _thread_local_end = ABSOLUTE(.);
     . = ALIGN(4);
   } GROUP_DATA_LINK_IN(RODATA_REGION, ROMABLE_REGION)
@@ -372,7 +374,7 @@ SECTIONS
 
 #if defined(CONFIG_ESP32_WIFI_IRAM_OPT)
     *libnet80211.a:( .wifi0iram  .wifi0iram.* .wifislpiram .wifislpiram.*)
-    *libpp.a:( .wifi0iram  .wifi0iram.* .wifislpiram .wifislpiram.*)
+    *libpp.a:( .wifi0iram  .wifi0iram.* .wifislpiram .wifislpiram.* .wifiorslpiram .wifiorslpiram.*)
 #endif
 
 #if defined(CONFIG_ESP32_WIFI_RX_IRAM_OPT)
@@ -491,7 +493,7 @@ SECTIONS
 
 #if !defined(CONFIG_ESP32_WIFI_IRAM_OPT)
     *libnet80211.a:( .wifi0iram  .wifi0iram.* .wifislpiram .wifislpiram.*)
-    *libpp.a:( .wifi0iram  .wifi0iram.* .wifislpiram .wifislpiram.*)
+    *libpp.a:( .wifi0iram  .wifi0iram.* .wifislpiram .wifislpiram.* .wifiorslpiram .wifiorslpiram.*)
 #endif
 
 #if !defined(CONFIG_ESP32_WIFI_RX_IRAM_OPT)

--- a/soc/xtensa/esp32/soc.c
+++ b/soc/xtensa/esp32/soc.c
@@ -26,6 +26,7 @@
 #include "soc/gpio_periph.h"
 #include "esp_spi_flash.h"
 #include "esp_err.h"
+#include "esp_timer.h"
 #include "esp32/spiram.h"
 #include "esp_app_format.h"
 #include <zephyr/sys/printk.h>
@@ -117,6 +118,8 @@ void __attribute__((section(".iram1"))) __esp_platform_start(void)
 	*wdt_rtc_protect = RTC_CNTL_WDT_WKEY_VALUE;
 	*wdt_rtc_reg &= ~RTC_CNTL_WDT_EN;
 	*wdt_rtc_protect = 0;
+
+	esp_timer_early_init();
 
 #if CONFIG_ESP32_NETWORK_CORE
 	/* start the esp32 network core before

--- a/soc/xtensa/esp32/soc.h
+++ b/soc/xtensa/esp32/soc.h
@@ -11,6 +11,7 @@
 #include <soc/soc_caps.h>
 #include <esp32/rom/ets_sys.h>
 #include <esp32/rom/spi_flash.h>
+#include <esp_rom_sys.h>
 
 #include <zephyr/types.h>
 #include <stdbool.h>

--- a/soc/xtensa/esp32s2/linker.ld
+++ b/soc/xtensa/esp32s2/linker.ld
@@ -175,6 +175,8 @@ SECTIONS
     *(.tdata.*)
     *(.tbss)
     *(.tbss.*)
+    *(.rodata_wlog)
+    *(.rodata_wlog*)
     _thread_local_end = ABSOLUTE(.);
     _rodata_reserved_end = ABSOLUTE(.);
     . = ALIGN(4);
@@ -274,7 +276,7 @@ SECTIONS
 
 #if defined(CONFIG_ESP32_WIFI_IRAM_OPT)
     *libnet80211.a:( .wifi0iram  .wifi0iram.* .wifislpiram .wifislpiram.*)
-    *libpp.a:( .wifi0iram  .wifi0iram.* .wifislpiram .wifislpiram.*)
+    *libpp.a:( .wifi0iram  .wifi0iram.* .wifislpiram .wifislpiram.* .wifiorslpiram .wifiorslpiram.*)
 #endif
 
 #if defined(CONFIG_ESP32_WIFI_RX_IRAM_OPT)
@@ -425,7 +427,7 @@ SECTIONS
 
 #if !defined(CONFIG_ESP32_WIFI_IRAM_OPT)
     *libnet80211.a:( .wifi0iram  .wifi0iram.* .wifislpiram .wifislpiram.*)
-    *libpp.a:( .wifi0iram  .wifi0iram.* .wifislpiram .wifislpiram.*)
+    *libpp.a:( .wifi0iram  .wifi0iram.* .wifislpiram .wifislpiram.* .wifiorslpiram .wifiorslpiram.*)
 #endif
 
 #if !defined(CONFIG_ESP32_WIFI_RX_IRAM_OPT)

--- a/soc/xtensa/esp32s2/soc.c
+++ b/soc/xtensa/esp32s2/soc.c
@@ -22,7 +22,10 @@
 #include "esp32s2/rom/cache.h"
 #include "soc/gpio_periph.h"
 #include "esp_spi_flash.h"
+#include "esp_cpu.h"
 #include "hal/cpu_ll.h"
+#include "hal/soc_ll.h"
+#include "esp_timer.h"
 #include "esp_err.h"
 #include "esp32s2/spiram.h"
 #include <zephyr/sys/printk.h>
@@ -91,6 +94,8 @@ void __attribute__((section(".iram1"))) __esp_platform_start(void)
 	*wdt_rtc_protect = RTC_CNTL_WDT_WKEY_VALUE;
 	*wdt_rtc_reg &= ~RTC_CNTL_WDT_EN;
 	*wdt_rtc_protect = 0;
+
+	esp_timer_early_init();
 
 #if CONFIG_ESP_SPIRAM
 
@@ -175,9 +180,10 @@ void IRAM_ATTR esp_restart_noos(void)
 
 	/* Reset wifi/ethernet/sdio (bb/mac) */
 	DPORT_SET_PERI_REG_MASK(DPORT_CORE_RST_EN_REG,
-				DPORT_BB_RST | DPORT_FE_RST | DPORT_MAC_RST |
-				DPORT_SDIO_RST | DPORT_SDIO_HOST_RST |
-				DPORT_EMAC_RST | DPORT_MACPWR_RST);
+				DPORT_BB_RST | DPORT_FE_RST | DPORT_MAC_RST | DPORT_BT_RST |
+				DPORT_BTMAC_RST | DPORT_SDIO_RST | DPORT_SDIO_RST |
+				DPORT_SDIO_HOST_RST | DPORT_EMAC_RST | DPORT_MACPWR_RST |
+				DPORT_RW_BTMAC_RST | DPORT_RW_BTLP_RST);
 	DPORT_REG_WRITE(DPORT_CORE_RST_EN_REG, 0);
 
 	/* Reset timer/spi/uart */
@@ -188,12 +194,9 @@ void IRAM_ATTR esp_restart_noos(void)
 		DPORT_UART_RST);
 	DPORT_REG_WRITE(DPORT_PERIP_RST_EN_REG, 0);
 
-	/* Set CPU back to XTAL source, no PLL, same as hard reset */
-	rtc_clk_cpu_freq_set_xtal();
-
 	/* Reset CPUs */
 	if (core_id == 0) {
-		SET_PERI_REG_MASK(RTC_CNTL_OPTIONS0_REG, RTC_CNTL_SW_PROCPU_RST_M);
+		soc_ll_reset_core(0);
 	}
 
 	while (true) {

--- a/soc/xtensa/esp32s2/soc.h
+++ b/soc/xtensa/esp32s2/soc.h
@@ -14,7 +14,8 @@
 #include <esp32s2/rom/ets_sys.h>
 #include <esp32s2/rom/spi_flash.h>
 #include <esp32s2/rom/cache.h>
-#include <esp_clk.h>
+#include <esp32s2/clk.h>
+#include <esp_rom_sys.h>
 
 #include <zephyr/types.h>
 #include <stdbool.h>

--- a/west.yml
+++ b/west.yml
@@ -62,7 +62,7 @@ manifest:
       groups:
         - hal
     - name: hal_espressif
-      revision: a85cf740929b7537e0bb07572b6c4f3456b04420
+      revision: 1e29548e4492807eb18cbaa77d3e78801cc7d07e
       path: modules/hal/espressif
       west-commands: west/west-commands.yml
       groups:

